### ISSUE TITLE
SSH connection cannot be established due to kex error

### ISSUE THREAD

**Issue by DocDriven** [CESNET/libnetconf2#234]
I have problems with the SSH connection for a netconf client and want to verify that I did not mess up anything.

My setup is netconf server (netopeer1) on board A, that can be accessed via the netopeer GUI. That lets me assume that the server is working fine.

On board B (client) I want to use the libnetconf2 lib to build a client. What I have read from the docs is that one needs to initialize the client first and then try to establish a SSH connection to the server.

```
#include "nc_client.h"

int
main(int argc, char* argv[])
{
    nc_client_init();

    const char *host = "172.27.14.66";
    uint16_t port = 0;
    struct nc_session *sess = NULL;

    sess = nc_connect_ssh(host, port, NULL);
    if(sess == NULL)
    {
        printf("Creating NETCONF session failed\n");
    }

    nc_client_destroy();

    return 0;
}
```
When trying to connect via SSH, I get an error stating:

`[ERR]: Starting the SSH session failed (kex error : no match for method server host key algo: server [ssh-rsa], client [ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256]).`

I can SSH from the command line to the server without any trouble; e.g. `ssh root@172.27.14.66` does work. So I do think that the SSH connection is fine as well. To make sure that sort of key is accepted, I generated all of them on the client via
`ssh-keygen -t rsa/dsa/ecdsa/ed25519`
and put the public keys under `/root/.ssh/authorized_keys`

I also made sure that I allowed root login in the `/etc/ssh/sshd_config` file and restart the ssh daemon afterwards.

I am not sure why this does not work. It would be helpful if you could confirm if I use the functions provided by libnetconf2 correctly. Also, if you have encountered this kind of problem before, please let me know. It is possible that I am on a completely wrong track and this is UNIX/libssh issue or something else.

Thanks!

**Comment by michalvasko** [comment]
Hi,
generally, the problem is what the error says meaning the server is providing an `ssh-rsa` algorithm host key but client accepts only `ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256` so it is simply a configuration issue. Based on what *libssh* version you are using, the allowed algorithms should be read from `/etc/ssh/ssh_config` (but hey are likely omitted there since you connected with `ssh(1)`), I think, or some default system ones are used.

So, in short, you can either change the server host key to use another algorithm or allow this algorithm for the client. For the former, you simply change the relevant *netopeer2* configuration (using *sysrepocfg*, for example) but for the latter I am not able to tell you the exact steps.

Regards,
Michal

**Comment by DocDriven** [comment]
Thanks Michal, I will look into that. It is good to know that the problem is just a configuration issue.

Sorry for posting this issue here, but I had to start somewhere.

Best regards!
